### PR TITLE
Feature/ranking

### PR DESCRIPTION
**Pull request by annajeannette**
Update di 9:00: tie handling aangepast, hij pakt nu het maximum (i.e. ranking van [0.1, 0.1, 0.1, 0.3] wordt [0.75, 0.75, 0.75, 1]. Nu klopt de functie verder wel.

---------------------------------

Shoot!

Ranking functie doet nu nog dingen die je niet helemaal verwacht, ga ik morgen mee verder! Het gaat dan om 'ties', die doet hij op zich goed, maar daardoor doet de interpolatie niet altijd wat je wil. Voor nu wel een PR zodat er alvast naar gekeken kan worden.

### CODE REVIEW

File: lir/transformers.py [hunk 1]
```diff
@@ -21,6 +23,41 @@ def transform(self, X):
...
+        self.rank_functions = [interp1d(X[:, i], ranks_X[:, i], bounds_error=False,
+                                        fill_value=(0, 1)) for i in range(X.shape[1])]
+        return self
+
+    def transform(self, X):
+        assert self.rank_functions
+        assert len(X.shape) == 2
+        assert X.shape[1] == len(self.rank_functions)
+        ranks = [self.rank_functions[i](X[:, i]) for i in range(X.shape[1])]
+        return np.asarray(ranks).transpose()
```
Comment (wowtor) [inline lir/transformers.py]: Met  `np.stack` ben je in één keer klaar, geen transpose nodig

File: lir/transformers.py [hunk 2]
```diff
@@ -21,6 +23,41 @@ def transform(self, X):
...
+
+    def fit(self, X, y=None):
+        assert len(X.shape) == 2
+        ranks_X = rankdata(X, axis=0)/len(X)
+        self.rank_functions = [interp1d(X[:, i], ranks_X[:, i], bounds_error=False,
+                                        fill_value=(0, 1)) for i in range(X.shape[1])]
+        return self
+
+    def transform(self, X):
+        assert self.rank_functions
```
Comment (wowtor) [inline lir/transformers.py]: wellicht een message meegeven, bijvoorbeeld
```
assert self.rank_functions, "transform() called before fit()"

File: lir/transformers.py [hunk 3]
```diff
@@ -21,6 +23,41 @@ def transform(self, X):
         return np.abs(X[:,:,0] - X[:,:,1])
 
 
+class RankTransformer(sklearn.base.TransformerMixin):
```
Comment (wowtor) [inline lir/transformers.py]: Naamsuggestie: `PercentileRankTransformer`, omdat hij naar range 0..1 gaat en niet 0..n

File: lir/transformers.py [hunk 4]
```diff
@@ -21,6 +23,41 @@ def transform(self, X):
...
+    """
+    Compute the rankings of a dataset, relative to another dataset.
+    Rankings are in [0, 1].
```
Comment (wowtor) [inline lir/transformers.py]: are in *range* [0, 1]
(current text may suggest set, not range)

File: lir/transformers.py (hunk 1 above)
Comment (annajeannette) [reply lir/transformers.py]: Als ik het uitprobeer heb ik daarna alsnog een transpose nodig

File: lir/transformers.py (hunk 2 above)
Comment (annajeannette) [reply lir/transformers.py]: Aangepast

File: lir/transformers.py (hunk 4 above)
Comment (annajeannette) [reply lir/transformers.py]: Aangepast

File: lir/transformers.py (hunk 3 above)
Comment (annajeannette) [reply lir/transformers.py]: Aangepast

**Comment by wowtor** [comment]
Stack geeft je de mogelijkheid om een axis te kiezen

Op di 15 mrt. 2022 09:02 schreef Jeannette ***@***.***>:

File: lir/transformers.py [hunk 5]
```diff
@@ -21,6 +23,41 @@ def transform(self, X):
         return np.abs(X[:,:,0] - X[:,:,1])
 
 
+class PercentileRankTransformer(sklearn.base.TransformerMixin):
+    """
+    Compute the rankings of a dataset, relative to another dataset.
```
Comment (vergep) [inline lir/transformers.py]:
```suggestion
    Compute the percentile rankings of a dataset, relative to another dataset.
```

File: lir/transformers.py [hunk 6]
```diff
@@ -21,6 +23,41 @@ def transform(self, X):
...
+    Rankings are in range [0, 1]. Handling ties: the maximum of the ranks that
```
Comment (vergep) [inline lir/transformers.py:29]: Is it possible to assign the median value? Never mind, it is completely arbitrary which rank is assigned. :)

File: lir/transformers.py [hunk 7]
```diff
@@ -21,6 +23,41 @@ def transform(self, X):
...
+    would have been assigned to all the tied values is assigned to each value.
+
+    To be able to compute the rankings of dataset Z relative to dataset X,
+    'fit' will create a ranking function for each feature, based on X.
+    'transform' will apply ranking of Z based on dataset X.
+
+    Transform:
+    Expects:
+        - X is of shape (n,f) with n=number of instances; f=number of features;
```
Comment (vergep) [inline lir/transformers.py]:
```suggestion
        - X is of shape (n,f) with n=number of measurements; f=number of features;
```

File: lir/transformers.py [hunk 8]
```diff
@@ -21,6 +23,41 @@ def transform(self, X):
...
+class PercentileRankTransformer(sklearn.base.TransformerMixin):
+    """
+    Compute the rankings of a dataset, relative to another dataset.
+    Rankings are in range [0, 1]. Handling ties: the maximum of the ranks that
+    would have been assigned to all the tied values is assigned to each value.
+
+    To be able to compute the rankings of dataset Z relative to dataset X,
+    'fit' will create a ranking function for each feature, based on X.
+    'transform' will apply ranking of Z based on dataset X.
+
```
Comment (vergep) [inline lir/transformers.py:35]: Fit:
Expects:
    - Z is of shape (m, f) with m= number of measurements, f=number of features

File: lir/transformers.py [hunk 9]
```diff
@@ -21,6 +23,41 @@ def transform(self, X):
...
+        - X is of shape (n,f) with n=number of instances; f=number of features;
+    Returns:
+        - rankings with shape (n, f)
+    """
+    def __init__(self):
+        self.rank_functions = None
+
+    def fit(self, X, y=None):
+        assert len(X.shape) == 2
+        ranks_X = rankdata(X, method='max', axis=0)/len(X)
```
Comment (vergep) [inline lir/transformers.py:53]: waarom niet 'method = average'?

File: lir/transformers.py [hunk 10]
```diff
@@ -21,6 +23,41 @@ def transform(self, X):
...
+        self.rank_functions = [interp1d(X[:, i], ranks_X[:, i], bounds_error=False,
+                                        fill_value=(0, 1)) for i in range(X.shape[1])]
+        return self
+
+    def transform(self, X):
+        assert self.rank_functions, "transform() called before fit()"
+        assert len(X.shape) == 2
+        assert X.shape[1] == len(self.rank_functions),\
+            "number of features used for fit() and transform()t should be equal"
```
Comment (vergep) [inline lir/transformers.py]:
```suggestion
            "number of features used for fit() and transform() should be equal"
```

**Review by vergep — APPROVED** [review]
Nice code! Looks good. Some minor comments.

File: lir/transformers.py (hunk 1 above)
Comment (annajeannette) [reply lir/transformers.py]: Alsnog stack toegevoegd

File: lir/transformers.py (hunk 5 above)
Comment (annajeannette) [reply lir/transformers.py]: Aangepast

File: lir/transformers.py (hunk 6 above)
Comment (annajeannette) [reply lir/transformers.py:29]: To make sure the interpolation works correctly, we need the maximum value. For example if you rank using the average instead of max, [0.1, 0.1, 0.1, 0.3] would lead to  [0.5, 0.5, 0.5, 1]. If you than interpolate, a value slightly larger than 0.1 would get a rank of slightly larger than 0.5, even though 75% of the values are below it!

File: lir/transformers.py (hunk 8 above)
Comment (annajeannette) [reply lir/transformers.py:35]: Aangepast (in aangepaste vorm, (n, f) gebruikt bij fit, en (n, f) bij transform)

File: lir/transformers.py (hunk 9 above)
Comment (annajeannette) [reply lir/transformers.py:53]: To make sure the interpolation works correctly, we need the maximum value. For example if you rank using the average instead of max, [0.1, 0.1, 0.1, 0.3] would lead to [0.5, 0.5, 0.5, 1]. If you than interpolate, a value slightly larger than 0.1 would get a rank of slightly larger than 0.5, even though 75% of the values are below it!

File: lir/transformers.py (hunk 8 above)
Comment (vergep) [reply lir/transformers.py:35]: Is het dan duidelijk dat n bij fit anders kan zijn dan n bij tansform?

File: lir/transformers.py (hunk 10 above)
Comment (annajeannette) [reply lir/transformers.py]: Aangepast

File: lir/transformers.py (hunk 8 above)
Comment (annajeannette) [reply lir/transformers.py:35]: Sorry, heb (n, f) gebruikt bij fit, en (m, f) bij transform.
